### ISSUE TITLE
":join" truncates a comments "<!--" in xml files

### ISSUE THREAD

**Issue by dtcyganok** [vim/vim#2494]
### Steps to reproduce

Create any xml file with content:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<foo>
  <bar/>
  <!-- comment -->
</foo>
```
Open this file with `vim -u NORC`
Move the cursor to the third line
Press "J"

Note: `:join` command works correct with `vim -u NONE`.

**Comment by AvianY** [comment]
I tried it and it works as it should with both NORC and NONE.

**Comment by vim-ml** [comment]
On 2017-12-24, AvianY wrote:
> I tried it and it works as it should with both NORC and NONE.

For me, with NONE and with

    vim -u NORC --cmd "set rtp=/usr/local/share/vim/vim80" comment.xml

it works fine.  (My plugin dependencies prevent me from using just
NORC.)  However, if I run just vim with no options and execute the
:join on line 3, I get this:

  <bar/> comment -->

So there is something common between Dmitrii's configuration and
mine that causes this, but I don't know what.  I also tried this,

    vim -u NORC --cmd "set rtp=/usr/local/share/vim/vim80" --cmd "filetype plugin indent on" --cmd "syntax on" comment.xml

but it also worked fine.  Is a puzzlement.  I'll poke around some
more.

Regards,
Gary

**Comment by vim-ml** [comment]
On 2017-12-24, Gary Johnson wrote:
> On 2017-12-24, AvianY wrote:
> > I tried it and it works as it should with both NORC and NONE.
> 
> For me, with NONE and with
> 
>     vim -u NORC --cmd "set rtp=/usr/local/share/vim/vim80" comment.xml
> 
> it works fine.  (My plugin dependencies prevent me from using just
> NORC.)  However, if I run just vim with no options and execute the
> :join on line 3, I get this:
> 
>   <bar/> comment -->
> 
> So there is something common between Dmitrii's configuration and
> mine that causes this, but I don't know what.  I also tried this,
> 
>     vim -u NORC --cmd "set rtp=/usr/local/share/vim/vim80" --cmd "filetype plugin indent on" --cmd "syntax on" comment.xml
> 
> but it also worked fine.  Is a puzzlement.  I'll poke around some
> more.

It happens when 'comments' is set to   

    s:

as it is for XML files, and 'formatoptions' includes 'j'.

You can reproduce the problem this way:

    $ vim -u NONE comment.xml
    :let comments = "s:"
    :set fo+=j
    :3join

where comment.xml contains Dmitrii's example.

Regards,
Gary

**Comment by tonymec** [comment]
I confirm the bug.

With my usual settings, where `'formatoptions'` is set to `croql` (and with the default settings from `ftplugin/xml.vim`) the comment is moved to after `<bar/>` with its `<!--` leader and one intervening space. If I undo this change, then `:setl fo+=j` and hit J again, the comment loses its `<!--` leader and there is one space between the ending `>` of `<bar/>` and the `c` of `comment`.

Using gvim 8.0.1424 (Huge) with GTK2-GNOME GUI.

Best regards,
Tony.

**Comment by tonymec** [comment]
The `j` in `'formatoptions'` means «When it makes sense, remove a comment leader when joining lines.» The question, of course, is «When does it make sense?» The examples in the help are for full-line `//` comments in C++ and IMHO, there, as well as e.g. `"` comments in vimscript, removing the comment leader **when joined to a previous line which is also a comment** makes sense. But in the case presented above by Dimitri, the comment is joined to a previous line **which is not a comment** (and does not end in a comment), and IMHO, here is doesn't "make sense" to remove a comment leader, regardless of the presence of `j` in `'formatoptions'

Best regards,
Tony.

**Comment by vim-ml** [comment]
On 2017-12-24, Gary Johnson wrote:
> On 2017-12-24, Gary Johnson wrote:
> > On 2017-12-24, AvianY wrote:
> > > I tried it and it works as it should with both NORC and NONE.
> > 
> > For me, with NONE and with
> > 
> >     vim -u NORC --cmd "set rtp=/usr/local/share/vim/vim80" comment.xml
> > 
> > it works fine.  (My plugin dependencies prevent me from using just
> > NORC.)  However, if I run just vim with no options and execute the
> > :join on line 3, I get this:
> > 
> >   <bar/> comment -->
> > 
> > So there is something common between Dmitrii's configuration and
> > mine that causes this, but I don't know what.  I also tried this,
> > 
> >     vim -u NORC --cmd "set rtp=/usr/local/share/vim/vim80" --cmd "filetype plugin indent on" --cmd "syntax on" comment.xml
> > 
> > but it also worked fine.  Is a puzzlement.  I'll poke around some
> > more.
> 
> It happens when 'comments' is set to   
> 
>     s:
> 
> as it is for XML files, and 'formatoptions' includes 'j'.
> 
> You can reproduce the problem this way:
> 
>     $ vim -u NONE comment.xml
>     :let comments = "s:"
>     :set fo+=j
>     :3join
> 
> where comment.xml contains Dmitrii's example.

The problem seems to be that get_last_leader_offset(), called at the
start of skip_comment(), sees the leading spaces in line 3 as the
middle part of the three-part comment defined in 'comments' and
identifies line 3 as a comment.  Consequently, when line 4 is joined
to line 3, the comment leader of line 4 is removed.

The html filetype has the same 'comments' value and suffers the
same problem.  I also discovered that the problem occurs when the
line on which the join is executed contains spaces anywhere, not
just at the start.

I don't understand the code and all the issues involved with
handling three-part comments well enough to have a solution, though.

Regards,
Gary

**Comment by tonymec** [comment]
I suppose that, ideally, a line headed by a middle-part comment header should not be regarded as a comment unless the (joined or not) line before that is either a first-part comment or (recursively) a middle-part comment preceded by etc. Also, a **first-** part comment header ought not to be suppressed unless it is joined to a preceding **end-** (not **middle-**) part comment line (whose end trailer is then also suppressed).

But I don't know if it is possible or worthwhile to program all this into "a simple (sic) text editor which is not a word processor".

Best regards,
Tony.

**Comment by chrisbra** [comment]
Agree, Vim should for middle part comments recursively for the next line above whether it is a start of a comment. This however might slow down Vim, as one needs to process many lines recursively. Perhaps it is better to not consider whitespace as comment leader for the first line to be joined. That is what https://github.com/chrisbra/vim/commit/44ec05f9a7070cca22a16bf0771faff2f9325af7 does.
